pkg/plugins/localfiles.go

```diff
@@ -10,76 +11,153 @@ import (
 	"github.com/grafana/grafana/pkg/util"
 )
 
-var _ fs.FS = (*LocalFS)(nil)
+var (
+	_ fs.File = &LocalFile{}
+
+	_ FS = &LocalFS{}
+	_ FS = &StaticFS{}
+)
 
 // LocalFS is a plugins.FS that allows accessing files on the local file system.
 type LocalFS struct {
-	// m is a map of relative file paths that can be accessed on the local filesystem.
-	// The path separator must be os-specific.
-	m map[string]*LocalFile
-
-	// basePath is the basePath that will be prepended to all the files (in m map) before accessing them.
+	// basePath is the basePath that will be prepended to all the files to get their absolute path.
 	basePath string
 }
 
-// NewLocalFS returns a new LocalFS that can access the specified files in the specified base path.
-// Both the map keys and basePath should use the os-specific path separator for Open() to work properly.
-func NewLocalFS(m map[string]struct{}, basePath string) LocalFS {
-	pfs := make(map[string]*LocalFile, len(m))
-	for k := range m {
-		pfs[k] = &LocalFile{
-			path: k,
+// NewLocalFS returns a new LocalFS that can access any file in the specified base path on the filesystem.
+// basePath must use os-specific path separator for Open() to work properly.
+func NewLocalFS(basePath string) LocalFS {
+	return LocalFS{basePath: basePath}
+}
+
+// fileIsAllowed takes an absolute path to a file and an os.FileInfo for that file, and it checks if access to that
+// file is allowed or not. Access to a file is allowed if the file is in the FS's Base() directory, and if it's a
+// symbolic link it should not end up outside the plugin's directory.
+func (f LocalFS) fileIsAllowed(basePath string, absolutePath string, info os.FileInfo) (bool, error) {
```

Comment:
To simplify things, I have moved `collectFilesWithin` that we use with the finder/loader into here and removed the ability to specify custom filepath.WalkerFunc on a LocalFS (from a previous iteration of this PR), because we don't really need that functionality for now